### ISSUE TITLE
Error: Collection "@nestjsplus/dyn-schematics" cannot be resolved.

### ISSUE THREAD

**Issue by AmrMonier** [nestjs/nest#9516]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current behavior

i'm trying to create a new module to integrate knex.js with nest following the instructions  from the [dev.to article](https://dev.to/nestjs/build-a-nestjs-module-for-knex-js-or-other-resource-based-libraries-in-5-minutes-12an?utm_source=dormosheio&utm_campaign=dormosheio) and when i run the following Command ``` nest g -c @nestjsplus/dyn-schematics dynpkg nest-knex``` the the following Error message appears
```
Error: Collection "@nestjsplus/dyn-schematics" cannot be resolved.
    at NodeModulesEngineHost.resolve (/media/monier/Data/Educational/Projects/Nest/fin-app/node_modules/@angular-devkit/schematics/tools/node-module-engine-host.js:75:19)
    at NodeModulesEngineHost._resolveCollectionPath (/media/monier/Data/Educational/Projects/Nest/fin-app/node_modules/@angular-devkit/schematics/tools/node-module-engine-host.js:80:37)
    at NodeModulesEngineHost.createCollectionDescription (/media/monier/Data/Educational/Projects/Nest/fin-app/node_modules/@angular-devkit/schematics/tools/file-system-engine-host-base.js:118:27)
    at SchematicEngine._createCollectionDescription (/media/monier/Data/Educational/Projects/Nest/fin-app/node_modules/@angular-devkit/schematics/src/engine/engine.js:162:40)
    at SchematicEngine.createCollection (/media/monier/Data/Educational/Projects/Nest/fin-app/node_modules/@angular-devkit/schematics/src/engine/engine.js:155:43)
    at NodeWorkflow.execute (/media/monier/Data/Educational/Projects/Nest/fin-app/node_modules/@angular-devkit/schematics/src/workflow/base.js:101:41)
    at main (/media/monier/Data/Educational/Projects/Nest/fin-app/node_modules/@angular-devkit/schematics-cli/bin/schematics.js:264:14)
    at Object.<anonymous> (/media/monier/Data/Educational/Projects/Nest/fin-app/node_modules/@angular-devkit/schematics-cli/bin/schematics.js:374:5)
    at Module._compile (node:internal/modules/cjs/loader:1105:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1159:10)

Failed to execute command: node @nestjsplus/dyn-schematics:dynpkg --name=nest-knex --dry-run --collection="@nestjsplus/dyn-schematics" --no-skipImport --language="ts" --sourceRoot="src" --spec
```

and i'm 100% sure that **@nestjsplus/dyn-schematics** is installed globally and this is the output  of ``` npm list -g ``` 
```
$ npm list -g
/home/monier/.nvm/versions/node/v18.1.0/lib
├── @nestjs/cli@8.2.5
├── @nestjsplus/dyn-schematics@1.0.12
├── corepack@0.10.0
└── npm@8.8.0

```

### Minimum reproduction code

a

### Steps to reproduce

1. create new nest project
2. npm install @nestjsplus/dyn-schematics -g
3. nest g -c @nestjsplus/dyn-schematics dynpkg nest-knex

### Expected behavior

it should create a new folder for the new module

### Package

- [ ] I don't know. Or some 3rd-party package
- [ ] <code>@nestjs/common</code>
- [ ] <code>@nestjs/core</code>
- [ ] <code>@nestjs/microservices</code>
- [ ] <code>@nestjs/platform-express</code>
- [ ] <code>@nestjs/platform-fastify</code>
- [ ] <code>@nestjs/platform-socket.io</code>
- [ ] <code>@nestjs/platform-ws</code>
- [ ] <code>@nestjs/testing</code>
- [ ] <code>@nestjs/websockets</code>
- [X] Other (see below)

### Other package

@nestjsplus/dyn-schematics@1.0.12

### Packages versions

```json
{
"dependencies": {
    "@nestjs/common": "^8.0.0",
    "@nestjs/core": "^8.0.0",
    "@nestjs/platform-express": "^8.0.0",
    "knex": "^2.0.0",
    "reflect-metadata": "^0.1.13",
    "rimraf": "^3.0.2",
    "rxjs": "^7.2.0"
  },
  "devDependencies": {
    "@nestjs/cli": "^8.0.0",
    "@nestjs/schematics": "^8.0.0",
    "@nestjs/testing": "^8.0.0",
    "@types/express": "^4.17.13",
    "@types/jest": "27.4.1",
    "@types/node": "^16.0.0",
    "@types/supertest": "^2.0.11",
    "@typescript-eslint/eslint-plugin": "^5.0.0",
    "@typescript-eslint/parser": "^5.0.0",
    "eslint": "^8.0.1",
    "eslint-config-prettier": "^8.3.0",
    "eslint-plugin-prettier": "^4.0.0",
    "jest": "^27.2.5",
    "prettier": "^2.3.2",
    "source-map-support": "^0.5.20",
    "supertest": "^6.1.3",
    "ts-jest": "^27.0.3",
    "ts-loader": "^9.2.3",
    "ts-node": "^10.0.0",
    "tsconfig-paths": "^3.10.1",
    "typescript": "^4.3.5"
  }
}
```

### Node.js version

18.1.0`

### In which operating systems have you tested?

- [ ] macOS
- [ ] Windows
- [X] Linux

**Comment by jmcdo29** [comment]
I was able to run the command in docker, but got a different error relating to the contents of the `@nestjsplus/dyn-schematics` scehmatic

<details>

<summary>You can see the commands executed here</summary>

```
❯ docker run  -it node sh
# npm i -g @nestjs/cli
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: '@angular-devkit/core@13.3.2',
npm WARN EBADENGINE   required: {
npm WARN EBADENGINE     node: '^12.20.0 || ^14.15.0 || >=16.10.0',
npm WARN EBADENGINE     npm: '^6.11.0 || ^7.5.6 || >=8.0.0',
npm WARN EBADENGINE     yarn: '>= 1.13.0'
npm WARN EBADENGINE   },
npm WARN EBADENGINE   current: { node: 'v15.9.0', npm: '7.5.3' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: '@angular-devkit/schematics@13.3.2',
npm WARN EBADENGINE   required: {
npm WARN EBADENGINE     node: '^12.20.0 || ^14.15.0 || >=16.10.0',
npm WARN EBADENGINE     npm: '^6.11.0 || ^7.5.6 || >=8.0.0',
npm WARN EBADENGINE     yarn: '>= 1.13.0'
npm WARN EBADENGINE   },
npm WARN EBADENGINE   current: { node: 'v15.9.0', npm: '7.5.3' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: '@angular-devkit/schematics-cli@13.3.2',
npm WARN EBADENGINE   required: {
npm WARN EBADENGINE     node: '^12.20.0 || ^14.15.0 || >=16.10.0',
npm WARN EBADENGINE     npm: '^6.11.0 || ^7.5.6 || >=8.0.0',
npm WARN EBADENGINE     yarn: '>= 1.13.0'
npm WARN EBADENGINE   },
npm WARN EBADENGINE   current: { node: 'v15.9.0', npm: '7.5.3' }
npm WARN EBADENGINE }

added 248 packages, and audited 249 packages in 11s

37 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
npm notice
npm notice New major version of npm available! 7.5.3 -> 8.8.0
npm notice Changelog: https://github.com/npm/cli/releases/tag/v8.8.0
npm notice Run npm install -g npm@8.8.0 to update!
npm notice
# npm i -g @nestjsplus/dyn-schematics

added 76 packages, and audited 77 packages in 4s

4 packages are looking for funding
  run `npm fund` for details

4 moderate severity vulnerabilities

To address issues that do not require attention, run:
  npm audit fix

Some issues need review, and may require choosing
a different dependency.

Run `npm audit` for details.
npm notice
npm notice New major version of npm available! 7.5.3 -> 8.8.0
npm notice Changelog: https://github.com/npm/cli/releases/tag/v8.8.0
npm notice Run npm install -g npm@8.8.0 to update!
npm notice
# nest g -c @nestjsplus/dyn-schematics dynpkg test
Error: NOT SUPPORTED: keyword "id", use "$id" for schema ID
    at Object.code (/usr/local/lib/node_modules/@nestjs/cli/node_modules/ajv/dist/vocabularies/core/id.js:6:15)
    at keywordCode (/usr/local/lib/node_modules/@nestjs/cli/node_modules/ajv/dist/compile/validate/index.js:454:13)
    at /usr/local/lib/node_modules/@nestjs/cli/node_modules/ajv/dist/compile/validate/index.js:222:17
    at CodeGen.code (/usr/local/lib/node_modules/@nestjs/cli/node_modules/ajv/dist/compile/codegen/index.js:439:13)
    at CodeGen.block (/usr/local/lib/node_modules/@nestjs/cli/node_modules/ajv/dist/compile/codegen/index.js:568:18)
    at iterateKeywords (/usr/local/lib/node_modules/@nestjs/cli/node_modules/ajv/dist/compile/validate/index.js:219:9)
    at groupKeywords (/usr/local/lib/node_modules/@nestjs/cli/node_modules/ajv/dist/compile/validate/index.js:208:13)
    at /usr/local/lib/node_modules/@nestjs/cli/node_modules/ajv/dist/compile/validate/index.js:192:13
    at CodeGen.code (/usr/local/lib/node_modules/@nestjs/cli/node_modules/ajv/dist/compile/codegen/index.js:439:13)
    at CodeGen.block (/usr/local/lib/node_modules/@nestjs/cli/node_modules/ajv/dist/compile/codegen/index.js:568:18)

Failed to execute command: node @nestjsplus/dyn-schematics:dynpkg --name=test --no-dry-run --collection="@nestjsplus/dyn-schematics" --no-skipImport --language="ts" --sourceRoot="src" --spec
```

</details>

The npm engine errors are expected. After `vim`ming the `@nestjsplsu/dyn-schematics/lib/schematics/dynpkg/schema.json` file to use `$id` instead of `id` everything worked as expected.

If I had to guess, you might be using a `nest` command installed via something other than npm. I just checked and having `@nestjsplus/dyn-schematics` installed via npm but `@nestjs/cli` installed via `yarn` causes a similar error to what you posted above.

**Comment by AmrMonier** [comment]
> 
@jmcdo29 
actually i don't even have yarn installed and i only use npm and i have installed it using nvm
